### DISCUSSION TITLE
Anyone added barcode functionality to the serenity framework in .net core?

### DISCUSSION THREAD

**Discussion by imhemayatsangin** [serenity-is/Serenity#6178] [Help] [answered]
Hello guys,
Please if anybody added barcode functionality to serenity application then please suggest me what to do and how to do?
I'm looking forward to hearing from you as soon as possible. Thanks

**Comment by sken78** [comment]


**Reply by imhemayatsangin** [reply]
Bro, i have products and i want to make a page or dialogue where i want to generate barcodes for each product and then print those barcodes. and also i want to scan those barcodes while selling products... i hope you have got what i mean. Thank you.

**Comment by sken78** [comment]


**Reply by imhemayatsangin** [reply]
I generated the barcode very easily and we can print it for each product. it's completed now i want to add each product to the order details grid by scanning the products' barcode and I'm stuck here if you can explain the process how to do?
 By the way Thanks for your response, appreciate it.

**Comment by sken78** [comment]


**Reply by imhemayatsangin** [reply]
Hahahaha Lol. I have those everything which you mentioned, I want to add a textbox instead of add button into the order detail grid and with the barcode scanner it should automatically add the same product and the data like price and quantity and show it to the order details grid. Anyway Thanks @sken78

**Comment by dfaruque** [comment, marked as answer]
Paste the following code into your OrderDetailsEditor's constructor
```
            let $barcodeInput = $('<input type="text" />');

            $barcodeInput.appendTo(this.toolbar.element)
                .keypress(event => {
                    if (event.keyCode == 13) {
                        let productId = $barcodeInput.val();

                        if (productId) {
                            let eqFilter = {};
                            eqFilter[ProductRow.Fields.ProductID] = productId

                            ProductService.List({ EqualityFilter: eqFilter, Take: 1 }, response => {
                                if (response.Entities.length > 0) {
                                    let pickedProduct = response.Entities[0];

                                    let orderDetails = this.value;
                                    orderDetails.push({
                                        ProductID: pickedProduct.ProductID,
                                        ProductName: pickedProduct.ProductName,
                                        UnitPrice: pickedProduct.UnitPrice,
                                        Quantity: 1,
                                        LineTotal: 1 * (pickedProduct.UnitPrice || 0)
                                    });

                                    this.value = orderDetails;
                                } else {
                                    Q.alert('Product not found!')
                                }
                            });
                        }
                    }
                });

```

**Reply by imhemayatsangin** [reply]
Thank you so much Faruque Ahmad bro, I will try it if any issue then will let you know. I send you an email as well regarding this. By the way thanks Bro for the response.

**Reply by imhemayatsangin** [reply]
Dear @dfaruque The code is working perfectly fine as i wanted thanks, but one issue, we must add one product to the orderdetails grid duplicates should not be allowed like how it is validated with the help of add product button. So how to check the productId with the rows product id to avoid duplication. here is my code:
`
 <reference path="../../Common/Helpers/GridEditorBase.ts" />
namespace SmartHMIS.Northwind {
    @Serenity.Decorators.registerClass()
    export class OrderDetailsEditor extends Common.GridEditorBase<OrderDetailRow> {
        protected getColumnsKey() { return "Northwind.OrderDetail"; }
        protected getDialogType() { return OrderDetailDialog; }
        protected getLocalTextPrefix() { return OrderDetailRow.localTextPrefix; }
        constructor(container: JQuery) {
            super(container);
            //for barcode scaner
            let $barcodeInput = $('<input type="text" />');
            $barcodeInput.appendTo(this.toolbar.element)
                .keypress(event => {
                    if (event.keyCode == 13) {
                        let productId = $barcodeInput.val();
                        if (productId) {
                            let eqFilter = {};
                            eqFilter[ProductRow.Fields.ProductID] = productId
                            ProductService.List({ EqualityFilter: eqFilter, Take: 1 }, response => {
                                if (response.Entities.length > 0) {
                                    let pickedProduct = response.Entities[0];
                                        let orderDetails = this.value;
                                        orderDetails.push({
                                            ProductID: pickedProduct.ProductID,
                                            ProductName: pickedProduct.ProductName,
                                            UnitPrice: pickedProduct.UnitPrice,
                                            Quantity: 1,
                                            Discount: 0,
                                            LineTotal: 1 * (pickedProduct.UnitPrice || 0) - 0
                                         
                                        });
                                   
                                    this.value = orderDetails;
                                } else {
                                    Q.alert('Product not found!')
                                }  
                            });
                        }
                    }
                }
                );
                //end of barcode scanner code
        }
        validateEntity(row, id) {
            row.ProductID = Q.toId(row.ProductID);   
            var sameProduct = Q.tryFirst(this.view.getItems(), x => x.ProductID === row.ProductID);
            if (sameProduct && this.id(sameProduct) !== id) {
                Q.alert('This product is already in order details!');
                return false;
            }
            row.ProductName = ProductRow.getLookup().itemById[row.ProductID].ProductName;
            row.LineTotal = (row.Quantity || 0) * (row.UnitPrice || 0) - (row.Discount || 0);
            return true;
        }
    }
}
`
waiting for your response brother. Thanks

**Reply by dfaruque** [reply]
```
            let $barcodeInput = $('<input type="text" />');

            $barcodeInput.appendTo(this.toolbar.element)
                .keypress(event => {
                    if (event.keyCode == 13) {
                        let productId = Q.toId($barcodeInput.val());
 
                        if (productId) {

                        var sameProduct = Q.tryFirst(this.view.getItems(), x => x.ProductID === productId );
                        if (sameProduct && this.id(sameProduct) !== id) {
                            Q.alert('This product is already in order details!');
                            return false;
                        }
                            let eqFilter = {};
                            eqFilter[ProductRow.Fields.ProductID] = productId

                            ProductService.List({ EqualityFilter: eqFilter, Take: 1 }, response => {
                                if (response.Entities.length > 0) {
                                    let pickedProduct = response.Entities[0];

                                    let orderDetails = this.value;
                                    orderDetails.push({
                                        ProductID: pickedProduct.ProductID,
                                        ProductName: pickedProduct.ProductName,
                                        UnitPrice: pickedProduct.UnitPrice,
                                        Quantity: 1,
                                        LineTotal: 1 * (pickedProduct.UnitPrice || 0)
                                    });

                                    this.value = orderDetails;
                                } else {
                                    Q.alert('Product not found!')
                                }
                            });
                        }
                    }
                });

```

**Reply by imhemayatsangin** [reply]
Thank you soo much dear bro @dfaruque , It's working fine. Appreciate it.
